[Alissa] iP

alissayarmantho opened the pull request:
# Saber Chatbot 
 Here are the tasks in your list:
 
 * [x]  Go to GitHub
 * [ ]  Download ~Duke~ Saber
 * [ ]  Manage your time better! :100: 
 
 > Saber has helped me manage my tasks. Definitely recommend!
 > -- A satisfied rando
 
 **Saber** is a _Chatbot_ designed for people who type fast and love the command-line. It features:
 
 * **Command-line interface** for keyboard-only navigation
 * **Adding** and **removing** tasks
 * Marking tasks as done
 * **Filtering** tasks by name
 
 With more features to come!
 
 Using Saber is simple:
 
 1. Click on [this jar link](https://github.com/alissayarmantho/ip/releases/tag/A-Jar)
 2. Download the `.jar` file
 3. Run the file
 4. Start adding tasks!
 
 Duke is a lightweight application written in Java. Here's the `main` method:
 
 ```java
 public static void main(String[] args) {
     new Saber("<filepath>/data/tasks.json").run();
 }
 ```

Commit:
Add Gradle support

alissayarmantho committed:
Add .class to gitignore

alissayarmantho committed:
Rename chatbot as Saber and add Saber logo

alissayarmantho committed:
Add Level 1 Requirements
- Add Command enums
- Add Greetings Message
- Add Goodbye Message

alissayarmantho committed:
Add Level 2 Requirements
- Add listMessage
- Add a String array with 100 elements to store tasks

Variables totalTask and newTaskIndex were added in the main function
as the number of tasks and the index at which new task is added to
are linked to each runtime

alissayarmantho committed:
Add Level-3 Requirements

- Add Command Class
- Add Task Class
- Add the ability to mark tasks as done
- Refactor main function:
        - To use Command Class
        - To use new functions in Saber Class
        - To use switch statement instead of if-else statement

alissayarmantho committed:
Add Level-4 Requirements
- Add Todo, Event, and Deadline class
- Rationale for hard coded + 4 in Command class
  is because time is indicated by '/by' or '/at' which
  are merely 4 chars. So + 4 is to remove the additional
  '/by ' or '/at ' substring from the time variable

alissayarmantho committed:
Fix initial test expected result

alissayarmantho committed:
Add Level-5 Requirements
- Add Custom Exception files
- Handle Exceptions
- Fix input parsing to be more strict

alissayarmantho committed:
Add Level-6 Requirements
- Add support for deleting tasks from the list
- Change the way of storing tasks, from using array to arraylist

alissayarmantho committed:
Remove trailing white spaces in print statements

alissayarmantho committed:
Add A-TextUiTesting
-Add Input and Expected output file to semi-automate the testing of
Saber

alissayarmantho committed:
Grammar fixes to print statements

alissayarmantho committed:
Fix parsing for todo and add

alissayarmantho committed:
Fix .gitignore (to ignore ACTUAL.TXT)

alissayarmantho committed:
Add dependency for saving .json

alissayarmantho committed:
Ignore task data folder when pushing to git

alissayarmantho committed:
Add new exceptions

alissayarmantho committed:
Add Level-7 Requirements
- Add TaskList class
- Add TaskStorage Class for saving and loading tasks from the hard
disk
- Adjust Task class and its children to take in isDone boolean field
- Refactor Saber.java to better organise the functions and to use all
the new updates around the codebase

alissayarmantho committed:
Merge branch 'branch-Level-7'
This merge will fulfill Level-7 Requirements

alissayarmantho committed:
Add time parser exception

alissayarmantho committed:
Enforce the use of Date and Time

alissayarmantho committed:
Merge branch 'branch-Level-8'
This merge will fulfill Level-8 Requirements

alissayarmantho committed:
Refactor date string handling + add more date format

alissayarmantho committed:
Make time parser and adjust the codebase accordingly

alissayarmantho committed:
Add A-MoreOOP Requirements
Make the codebase more OOP:
- Extract all Command classes
- Extract all UI classes (related to the Command)
- Refactor old Command class to become Parser class instead
- Adjust codebase to fit these major changes

alissayarmantho committed:
Add A-Packages Requirement

alissayarmantho committed:
Add A-Jar Requirement

alissayarmantho committed:
Minor refactor of parsing logic

alissayarmantho committed:
Add Level-9 Requirements
Add a way for user to find tasks with the given keyword

alissayarmantho committed:
Merge branch 'branch-Level-9'
This merge will fulfill Level-9 Requirements

alissayarmantho committed:
Remove unused variable in ListUI

alissayarmantho committed:
Add A-JavaDoc Requirements

alissayarmantho committed:
Merge branch 'branch-A-JavaDoc'
This merge will fulfill A-JavaDoc Requirement

alissayarmantho committed:
Add A-CodingStandard Requirements

alissayarmantho committed:
Merge branch 'branch-A-CodingStandard'
This merge will fulfill A-CodingStandard Requirements

alissayarmantho committed:
Add A-JUnit Requirements
-Add test methods for ByeCommand class
-Add test methods for TaskList class

alissayarmantho committed:
Fix Todo UI message output

alissayarmantho committed:
Fix Text-UI test

alissayarmantho committed:
Ignore text-ui-test/sources.txt

alissayarmantho committed:
Remove unused print statement in SaberParser.java

alissayarmantho committed:
Remove generated javadocs files

gycgabriel commented on src/main/java/saber/Saber.java:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    /***
+     * Saber constructor. Takes in a filepath for TaskStorage to know which file
```
May want to start javadocs for methods using a verb in action form, _e.g. Returns, Takes_ instead.

alissayarmantho replied:
Ok

gycgabriel commented on src/main/java/saber/Saber.java:
```diff
@@ -0,0 +1,136 @@
+ */
+public class Saber {
+
+    /** The storage of the Saber application that handles the loading and saving of the tasks to
+     * the hard disk. Tasks are only loaded from hard disk at the beginning of the application and saved to the hard
+     * disk when the application ends */
```
I like how you wrote comments for variable names 👍. The comment seems rather long for a variable, perhaps the information of when storing is done could be better placed under the class comment for Saber and Storage class.

alissayarmantho replied:
I see.

gycgabriel commented on src/main/java/saber/Saber.java:
```diff
@@ -0,0 +1,136 @@
+package saber;
+
+import saber.commands.SaberCommand;
+
+import saber.exceptions.SaberCommandNotFoundException;
+import saber.exceptions.SaberStorageLoadException;
+import saber.exceptions.SaberStorageStoreException;
+
+import saber.parser.SaberParser;
+import saber.storage.TaskStorage;
+import saber.task.Deadline;
+import saber.task.Event;
+import saber.task.Todo;
+
+import saber.tasklist.TaskList;
+import saber.ui.SaberUI;
```
Maybe these import statements are redundant as the classes are under the same package. Any reason for importing these? Same for other files.

alissayarmantho replied:
I think they are considered different packages.

gycgabriel commented on src/main/java/saber/Saber.java:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    /***
+     * The main function
+     * @param args the arguments of the main function when running the program
```
Might have forgotten to add an empty line between description and parameter section, in several other places as well.
```suggestion
     * The main function
     
     * @param args the arguments of the main function when running the program
```

alissayarmantho replied:
Ok. Funnily enough Intellij auto javadocs generator doesn't seem to follow our module's coding standard...

gycgabriel commented on src/main/java/saber/parser/SaberParser.java:
```diff
@@ -0,0 +1,175 @@
+import saber.exceptions.MissingTimeException;
+import saber.exceptions.SaberCommandNotFoundException;
+
+/** A class that encapsulates a parser for the Saber application (to parse user commands) */
```
Class javadocs should start on a new line after `/**`
```suggestion
/** 
 * A class that encapsulates a parser for the Saber application (to parse user commands) 
 */
```

alissayarmantho replied:
Ok

gycgabriel commented on src/main/java/saber/Saber.java:
```diff
@@ -0,0 +1,136 @@
+
+import java.util.ArrayList;
+
+/***
+ *  A class that represents the Saber application
+ */
```
May want to start javadoc comments using two asterisks only, and end javadocs with fullstop. I noticed this issue in several other places too.
```suggestion
/**
 *  A class that represents the Saber application.
 */
```

alissayarmantho replied:
Ok

gycgabriel commented on src/main/java/saber/commands/AddCommand.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class AddCommand extends SaberCommand {
+    private Task task;
+    private boolean isMissingDescription;
```
I like the naming of this boolean and other variables in several other places 👍.

gycgabriel commented on src/main/java/saber/commands/AddCommand.java:
```diff
@@ -0,0 +1,49 @@
+     * A function to determine whether the current command is a terminating command (a ByeCommand)
+     * @return false
+     */
+    public boolean isExit() {
```
Would naming this checkExit or getIsExit be better because method names should be action verbs?

gycgabriel replied:
Unrelated (purely optional) suggestion: if all commands have this property, maybe can consider converting isExit into a variable and have one getter function in the parent class Command that all command types inherit?

alissayarmantho replied:
Now that I think about it, I could've made this inside SaberCommand.java and do an instanceof instead, so there is no need for each children to define its own isExit()

BuzzingBuzzard replied:
I agree with Gabriel's suggestion. Since the isExit() function only deals with one property (true/false for whether the program is running), and every command has this property, perhaps it's better to convert it into a variable?

gycgabriel commented on src/main/java/saber/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,69 @@
+import saber.tasklist.TaskList;
+import saber.ui.DeadlineUI;
+
+/**
```
Good javadoc syntax here, slashes done correctly here and in several other places 👍.

alissayarmantho replied:
Thanks

gycgabriel commented on src/main/java/saber/commands/EventCommand.java:
```diff
@@ -0,0 +1,69 @@
+     * @param isMissingDescription whether the event description is missing in the command
+     * @param isMissingTime whether the event time is missing in the command
+     */
+    public EventCommand(String eventTask,
+                            String eventTime,
+                            boolean isMissingDescription,
+                            boolean isMissingTime) {
```
Good line break in arguments to make it more readable 👍.
Indentation for wrapped lines should be 8 spaces more than the parent line though, so perhaps like this:
```suggestion
    public EventCommand(String eventTask,
            String eventTime,
            boolean isMissingDescription,
            boolean isMissingTime) {
```

alissayarmantho replied:
Ok

gycgabriel commented on src/main/java/saber/parser/SaberParser.java:
```diff
@@ -0,0 +1,175 @@
+
+/** Import all saber commands as the parse function require all of them */
+import saber.Saber;
+import saber.commands.*;
```
Should we be using * operator in our imports? 
According to the coding standard, imported classes should always be listed explicitly.

alissayarmantho replied:
Ah yes.

gycgabriel commented on src/main/java/saber/storage/TaskStorage.java:
```diff
@@ -0,0 +1,186 @@
+package saber.storage;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+
+import java.nio.file.InvalidPathException;
+
+import java.util.ArrayList;
+
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+
+import saber.Saber;
+import saber.tasklist.TaskList;
+import saber.exceptions.SaberStorageLoadException;
+import saber.exceptions.SaberStorageStoreException;
+import saber.exceptions.SaberTaskTypeNotFoundException;
+
+import saber.task.Deadline;
+import saber.task.Event;
+import saber.task.Task;
+import saber.task.Todo;
```
Neatly organised imports even when there are many things to import! 👍 👍

alissayarmantho replied:
Thanks

gycgabriel commented on src/main/java/saber/storage/TaskStorage.java:
```diff
@@ -0,0 +1,186 @@
+     * @throws SaberStorageLoadException if there is any error in parsing or IO
+     */
+    @SuppressWarnings("unchecked")
+    // This method needs to suppress unchecked warnings due to the nature of the conversion
+    // of JSON Object to saber.task.Task
+    // This method will still be type safe as only tasks are stored in the file
```
(Unrelated) That's cool to use JSON 😃 👍  !

alissayarmantho replied:
:)

gycgabriel commented on src/main/java/saber/Saber.java:
```diff
@@ -0,0 +1,136 @@
+    /***
+     * Saber constructor. Takes in a filepath for TaskStorage to know which file
+     * to load the tasks from to TaskList and to save to from the TaskList
+     * @param filePath the file path where the tasks will be saved to (the tasks will be saved at
+     *                 filePath/data/tasks.json)
```
Good details in javadocs that informs readers how the filePath argument is used 👍.

alissayarmantho replied:
Thanks !!!

gycgabriel commented on src/main/java/saber/storage/TaskStorage.java:
```diff
@@ -0,0 +1,186 @@
+            java.nio.file.Path path = java.nio.file.Paths.get(this.path, "data", "tasks.json");
+            File file = new File(path.toString());
+
+            // Suppress inspection for these methods because we want to ignore the result of these methods as in a
+            // typical run these methods will not give any error as the path we are passing in is not a user-defined
+            // path. As such, it is safe to suppress these warnings
```
Good substantiation of why suppress is used here 👍.

alissayarmantho replied:
Thanks!

gycgabriel commented on src/main/java/saber/storage/TaskStorage.java:
```diff
@@ -0,0 +1,186 @@
+            java.nio.file.Path path = java.nio.file.Paths.get(this.path, "data", "tasks.json");
+            File file = new File(path.toString());
+
+            // Suppress inspection for these methods because we want to ignore the result of these methods as in a
+            // typical run these methods will not give any error as the path we are passing in is not a user-defined
+            // path. As such, it is safe to suppress these warnings
```
(Unrelated to coding standard) I'm really not sure, would it be better for @SuppressWarnings to be placed here instead of above the entire method?

alissayarmantho replied:
Hmm... I think I could have done that, but then there are a lot of casting done here so I just put it on top of the method haha

gycgabriel commented on src/main/java/saber/time/SaberTime.java:
```diff
@@ -0,0 +1,97 @@
+            DateTimeFormatter.ofPattern("MMM, d yyyy HH:mm"),
+            DateTimeFormatter.ofPattern("d MMMM yyyy HH:mm"),
+            DateTimeFormatter.ofPattern("d MMM yyyy HH:mm"),
+    };
```
hey, I like how clean this bit of code is 👍

gycgabriel commented on src/main/java/saber/ui/DeadlineUI.java:
```diff
@@ -0,0 +1,75 @@
+public class DeadlineUI extends SaberCommandUI {
+    private String successMessage;
+
+    protected final String missingDescriptionError = "      I'm really sorry, Master.\n"
+            + "      I'm ... not sure what task you want me\n"
+            + "      to add ...\n";
+
+    protected final String missingTimeError = "      I'm sorry, Master.\n"
+            + "      I won't be able to ensure that you\n"
+            + "      do the task on time without\n"
+            + "      knowing the deadline...\n";
+
+    protected final String parsingTimeError = "      I'm sorry, Master. I'm unable\n"
+            + "      to understand the date you\n"
+            + "      just told me. Maybe..., Master\n"
+            + "      can try to use other format?\n";
```
Neat! 👍

gycgabriel commented on src/main/java/saber/ui/SaberUI.java:
```diff
@@ -0,0 +1,111 @@
+            +  "      know...I won't remember our\n"
+            +  "      interaction either. A new\n"
+            +  "      saber.Saber will serve you instead.\n\n"
+            +  "      Have I...done well, Master?\n";
```
(Unrelated to coding standard)
💟 good customization 💟

alissayarmantho replied:
(Unrelated), a lot of my friends laugh at this :'))))))

gycgabriel commented on src/main/java/saber/ui/SaberUI.java:
```diff
@@ -0,0 +1,111 @@
+public class SaberUI {
+    private Scanner input = new Scanner(System.in);
+
+    protected static final String logo = "\n                            __,    _,  _ __  ______ _ __\n" +
+            "              />           (      / | ( /  )(  /   ( /  )\n" +
+            "             /<             `.   /--|  /--<   /--   /--<\n" +
+            "            /<            (___)_/   |_/___/ (/____//   \\_\n" +
+            "           /<\n" +
+            " |\\_______{o}----------------------------------------------------------_\n" +
+            "[\\\\\\\\\\\\\\\\\\\\\\{*}:::<=============================================-       >\n" +
+            " |/~~~~~~~{o}----------------------------------------------------------~\n" +
+            "           \\<\n" +
+            "            \\<\n" +
+            "             \\>\n";
+
+    protected static final String lineBreak = "      |                                     |\n" +
+            "{o)xxx|===============- * -===============|xxx(o}\n" +
+            "      |                                     |\n";
+
+    protected final String greeting = "      I hear your command, Master.\n" +
+            "      What can I do for you?\n";
```
Might have to add the `+` in front of each line according to the coding standard. I wish there was a way to do this automatically...

alissayarmantho replied:
Ok

gycgabriel reviewed (commented):
I found your code to follow the coding standard very well besides a few minor violations. 

It was fun reading your code as it was very readable with comments to guide the reader. It was also a learning experience for me as I learnt about using JSON to store and load which I had not thought of 👍 😃

BuzzingBuzzard commented on src/main/java/saber/exceptions/SaberException.java:
```diff
@@ -0,0 +1,15 @@
+ * A custom exception class to indicate all Saber-related error (during runtime)
+ */
+public class SaberException extends RuntimeException {
+
```
It's interesting that you let SaberException extend RuntimeException. Please correct me if I'm wrong, yet I thought that RuntimeException are unchecked exceptions, which happen because of bad programming / programmers' fault. On the other hand, exceptions that occur due to user's faults (like invalid input, file not found etc.) are checked exceptions (which can happen even if the programmer did everything correctly).

I thought that what SaberException deals with is more like checked exceptions, where you programmed perfectly, yet users may do silly things like missing arguments. 

Just a thought. Maybe you can give more clarifications?

BuzzingBuzzard commented on src/main/java/saber/exceptions/SaberException.java:
```diff
@@ -0,0 +1,15 @@
+    public SaberException(String message) {
+        super(message);
+    }
+}
```
I like how you considered all possible exceptions, and listed them out as classes extending SaberException. However, as an optional personal opinion, I see that each subclass exception only uses the super constructor, and doesn't have its own special behavior; thus perhaps you can replace all subclass exceptions with SaberException itself, and just type in different (specialized) error messages? 

In this way, you only need to catch one big SaberException at the end of everything, and print out its specialized error message, which would tell the user what's wrong all the same. Then you don't need to have lots of different exception classes, and don't need to have catch block for them all. Could make your code clearer.

Just a personal opinion though. There's nothing wrong listing out each exception as a subclass.

BuzzingBuzzard commented on src/main/java/saber/ui/SaberUI.java:
```diff
@@ -0,0 +1,111 @@
+            "      |                                     |\n";
+
+    protected final String greeting = "      I hear your command, Master.\n" +
+            "      What can I do for you?\n";
```
Coolest logo & line break every made! The Force is strong with this one :)

BuzzingBuzzard reviewed (commented):
Overall, great code quality! I can see you put in a lot of efforts structuring the code, and even styling the displays. See individual comments for suggestions. Keep up the good work :D

alissayarmantho committed:
Add checkstyle

alissayarmantho committed:
Merge branch 'add-gradle-support'
Fulfill A-Gradle requirements

alissayarmantho committed:
Add JSON Simple dependancies

alissayarmantho committed:
Fix checkstyle errors

alissayarmantho committed:
Change name in build.gradle

alissayarmantho committed:
Add javafx to gradle

alissayarmantho committed:
Fix mainclassname for build.gradle

alissayarmantho committed:
Add Level-10 Requirements

alissayarmantho committed:
Merge branch 'branch-Level-10'
Fulfill Level-10 Requirements

alissayarmantho committed:
Increase width of window

alissayarmantho committed:
Fix JavaFx bugs

alissayarmantho committed:
Fix javafx closing method

alissayarmantho committed:
Fix user experience when exiting Saber

alissayarmantho committed:
Add A-Assertions
No assertion

There is a need to do assertion to verify assumptions invariants

Add assert to Done and Delete Command

This is done in order to ensure that when the command proceeds
the index that it is trying to get is always within the taskList
available

Other changes:
Add javadoc to main method in Launcher.java

alissayarmantho committed:
Merge pull request #1 from alissayarmantho/branch-A-Assertions

Branch A- Assertions

alissayarmantho committed:
Add A-CI
No CI

CI is needed to ensure that the code pushed to master branch is of a
certain standard

Add CI

Done by adding gradle.yml to .github/workflows

alissayarmantho committed:
Fix CI not passing
Unable to pass CI

Fix the CI errors

Done in order to pass CI

alissayarmantho committed:
Fix checkstyle error

alissayarmantho committed:
Remove text redirection CI test

alissayarmantho committed:
Remove windows CI check

alissayarmantho committed:
Merge pull request #2 from alissayarmantho/branch-A-CI

Branch A-CI

alissayarmantho committed:
Add sorting ability by done-ness

alissayarmantho committed:
Merge branch 'branch-C-Sort'
Fulfill C-Sort, BCD-Extension Requirements

alissayarmantho committed:
Fix javadocs to fit the coding standard

alissayarmantho committed:
Merge pull request #3 from alissayarmantho/branch-A-CodeQuality

Branch A-CodeQuality

alissayarmantho committed:
Set theme jekyll-theme-cayman

alissayarmantho committed:
Add User Guide

alissayarmantho committed:
Add full-stop to the user guide

alissayarmantho committed:
Add v0.2 release

alissayarmantho committed:
Show storage loading and storing errors

alissayarmantho commented:
Class has ended for this module
